Update: (Indexer) BinSearch to support RSS and basic group guessing.

Nyuels opened the pull request:
#### Database Migration
NO

#### Description
This PR introduces RSS/empty searches to the BinSearch Indexer.
To allow this, two new advanced settings are added which make use of BinSearch's [group browsing](https://binsearch.info/browse.php):

- `Fallback Newsgroup`: specifies the group to browse in case of RSS/empty searches
- `Fallback Server`: specifies the backend server BinSearch will use to browse the `Fallback Newsgroup` (_Primary_ indexes the most popular groups, _Secondary_ other groups)

In addition, this PR would add simple category guessing based on the list of filetypes BinSearch returns for each result entry.

#### Screenshot (if UI related)

#### Todos
- [ ] Extend the basic category guessing.
- [ ] Split PR into the two features?

#### Issues Fixed or Closed by this PR

Qstick commented on src/NzbDrone.Core/Indexers/Definitions/BinSearch.cs:
```diff
@@ -52,7 +56,7 @@ private IndexerCapabilities SetCapabilities()
                 TvSearchParams = new List<TvSearchParam>
                                    {
-                                       TvSearchParam.Q, TvSearchParam.Season, TvSearchParam.Ep
+                                       TvSearchParam.Q
```
We should keep the Season and Ep as we use it to concat the search term for TV

Qstick commented on src/NzbDrone.Core/Indexers/Definitions/BinSearch.cs:
```diff
@@ -165,65 +198,141 @@ public IList<ReleaseInfo> ParseResponse(IndexerResponse indexerResponse)
+                    // Guess category based on filetype
+                    if (fileTypes.Any(fileType => CommonVideoFileTypes.Contains(fileType)))
+                    {
+                        release.Categories = new List<IndexerCategory> { NewznabStandardCategory.Movies, NewznabStandardCategory.TV, NewznabStandardCategory.XXX };
```
This is problematic, we don't want to tag everything with XXX

Nyuels replied:
Would Movies and TV category be okay though?
Otherwise it is probably better to wait for a uniform implementation (https://github.com/Prowlarr/Prowlarr/issues/394).

Qstick replied:
That's probably fine for now.

Qstick commented on src/NzbDrone.Core/Indexers/Definitions/BinSearch.cs:
```diff
@@ -165,65 +198,141 @@ public IList<ReleaseInfo> ParseResponse(IndexerResponse indexerResponse)
+        [FieldDefinition(2, Label = "Fallback Newsgroup", Advanced = true, Type = FieldType.Textbox, HelpText = "Use this newsgroup for empty searches.")]
+        public string FallbackGroup { get; set; }
+
+        [FieldDefinition(3, Label = "Fallback Server", Advanced = true, Type = FieldType.Select, SelectOptions = typeof(BinSearchFallbackServer), HelpText = "Use this Binsearch server for empty searches.")]
```
Is this really needed? What's the difference when selecting Primary or Secondary?

Nyuels replied:
I haven't checked all, but from comparisons I deducted that the list of indexed groups for [Primary](https://binsearch.info/groups.php) and [Secondary](https://binsearch.info/groups.php?server=2) is disjunct:
Primary "contains" the most popular groups, while Secondary "contains" the remaining ones, BinSearch indexes.

Qstick commented on src/NzbDrone.Core/Indexers/Definitions/BinSearch.cs:
```diff
@@ -165,65 +198,141 @@ public IList<ReleaseInfo> ParseResponse(IndexerResponse indexerResponse)
+    {
+        public BinSearchSettingsValidator()
+        {
+            RuleFor(c => c.BaseUrl).NotEmpty();
```
We default to the first URL when empty, so this really isn't needed.

Qstick commented:
Feel free to bring this out of draft when ready

Nyuels commented:
> Feel free to bring this out of draft when ready

Will do, I wasn't quite happy about the category tagging, so I tried to make it more reusable.
It would probably be a good idea to update/overhaul/expand on that in the future, depending on needs.

Nyuels commented:
Updated the category guessing to also include special token in the release title.
By doing so I moved both Parsers into Prowlarr.Core.Parser.Category for future reusability (e.g. https://github.com/Prowlarr/Prowlarr/issues/394).

Qstick commented:
> Updated the category guessing to also include special token in the release title.
> By doing so I moved both Parsers into Prowlarr.Core.Parser.Category for future reusability (e.g. #394).

Nice work 👍 I’ll have a go
